File: app/src/main/java/com/dumuzeyn/mp3player/Track.java
package com.dumuzeyn.mp3player;

import android.net.Uri;
import java.util.Locale;

public class Track {
    public final String album;
    public final String albumArtist;
    public final String artist;
    public final long dateAdded;
    public final int discNumber;
    public final int durationMs;
    public final String genre;
    public final String title;
    public final String trackId;
    public final String uri;
    public final long fileSize;
    public final long lastModified;
    public final long lastCompletedAt;
    public final long lastPlayedAt;
    public final int playCount;
    public final int skipCount;
    public final int trackNumber;
    public final int year;
    public final String fingerprint;
    public final String normalizedSearchText;

    public Track(String uri, String title, String artist) {
        this(uri, title, artist, "Unknown album", "Unknown genre", 0);
    }

    public Track(String uri, String title, String artist, String album, String genre) {
        this(uri, title, artist, album, genre, 0);
    }

    public Track(String uri, String title, String artist, String album, String genre, int durationMs) {
        this(TrackIdentity.fromLegacyUri(uri), uri, title, artist, album, genre, durationMs,
                -1L, 0L, "");
    }

    public Track(String trackId, String uri, String title, String artist, String album,
            String genre, int durationMs, long fileSize, long lastModified, String fingerprint) {
        this(trackId, uri, title, artist, album, artist, genre, 0, 0, 0, durationMs,
                fileSize, lastModified, fingerprint, 0, 0, System.currentTimeMillis(), 0L, 0L);
    }

    public Track(String trackId, String uri, String title, String artist, String album,
            String albumArtist, String genre, int year, int trackNumber, int discNumber,
            int durationMs, long fileSize, long lastModified, String fingerprint,
            int playCount, int skipCount, long dateAdded, long lastPlayedAt,
            long lastCompletedAt) {
        this.trackId = trackId == null || trackId.trim().isEmpty()
                ? TrackIdentity.create() : trackId;
        this.uri = uri;
        this.title = title;
        this.artist = artist;
        this.album = album;
        this.albumArtist = albumArtist == null || albumArtist.trim().isEmpty()
                ? artist : albumArtist;
        this.genre = GenreNormalizer.normalize(genre);
        this.year = Math.max(0, year);
        this.trackNumber = Math.max(0, trackNumber);
        this.discNumber = Math.max(0, discNumber);
        this.durationMs = Math.max(0, durationMs);
        this.fileSize = fileSize;
        this.lastModified = Math.max(0L, lastModified);
        this.fingerprint = fingerprint == null ? "" : fingerprint;
        this.playCount = Math.max(0, playCount);
        this.skipCount = Math.max(0, skipCount);
        this.dateAdded = Math.max(0L, dateAdded);
        this.lastPlayedAt = Math.max(0L, lastPlayedAt);
        this.lastCompletedAt = Math.max(0L, lastCompletedAt);
        this.normalizedSearchText = buildSearchText();
    }

    public Track withLocation(String newUri, long newSize, long newLastModified,
            String newFingerprint) {
        return new Track(trackId, newUri, title, artist, album, genre, durationMs,
                newSize, newLastModified, newFingerprint).withDetails(this);
    }

    public Track withMetadata(String newTitle, String newArtist, String newAlbum,
            String newAlbumArtist, String newGenre, int newYear, int newTrackNumber,
            int newDiscNumber) {
        return new Track(trackId, uri, newTitle, newArtist, newAlbum, newAlbumArtist,
                newGenre, newYear, newTrackNumber, newDiscNumber, durationMs, fileSize,
                lastModified, fingerprint, playCount, skipCount, dateAdded, lastPlayedAt,
                lastCompletedAt);
    }

    public Track withPlaybackStats(int newPlayCount, int newSkipCount, long newLastPlayedAt,
            long newLastCompletedAt) {
        return new Track(trackId, uri, title, artist, album, albumArtist, genre, year,
                trackNumber, discNumber, durationMs, fileSize, lastModified, fingerprint,
                newPlayCount, newSkipCount, dateAdded, newLastPlayedAt, newLastCompletedAt);
    }

    private Track withDetails(Track source) {
        return new Track(trackId, uri, title, artist, album, source.albumArtist, genre,
                source.year, source.trackNumber, source.discNumber, durationMs, fileSize,
                lastModified, fingerprint, source.playCount, source.skipCount,
                source.dateAdded, source.lastPlayedAt, source.lastCompletedAt);
    }

    public Uri asUri() {
        return Uri.parse(this.uri);
    }

    static String normalizeSearchText(String value) {
        if (value == null || value.isEmpty()) {
            return "";
        }
        return value.toLowerCase(Locale.ROOT).replaceAll("\\s+", " ").trim();
    }

    private String buildSearchText() {
        StringBuilder value = new StringBuilder()
                .append(title).append(' ').append(artist).append(' ')
                .append(album).append(' ').append(genre);
        if (!normalizeSearchText(albumArtist).equals(normalizeSearchText(artist))) {
            value.append(' ').append(albumArtist);
        }
        if (year > 0) {
            value.append(' ').append(year);
        }
        return normalizeSearchText(value.toString());
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/Track.kt
package com.dumuzeyn.mp3player

import android.net.Uri
import java.util.Locale

class Track(
    trackId: String?,
    @JvmField val uri: String,
    @JvmField val title: String,
    @JvmField val artist: String,
    @JvmField val album: String,
    albumArtist: String?,
    genre: String?,
    year: Int,
    trackNumber: Int,
    discNumber: Int,
    durationMs: Int,
    @JvmField val fileSize: Long,
    lastModified: Long,
    fingerprint: String?,
    playCount: Int,
    skipCount: Int,
    dateAdded: Long,
    lastPlayedAt: Long,
    lastCompletedAt: Long,
) {
    @JvmField val trackId: String = trackId?.takeIf { it.isNotBlank() } ?: TrackIdentity.create()
    @JvmField val albumArtist: String = albumArtist?.takeIf { it.isNotBlank() } ?: artist
    @JvmField val genre: String = GenreNormalizer.normalize(genre)
    @JvmField val year: Int = year.coerceAtLeast(0)
    @JvmField val trackNumber: Int = trackNumber.coerceAtLeast(0)
    @JvmField val discNumber: Int = discNumber.coerceAtLeast(0)
    @JvmField val durationMs: Int = durationMs.coerceAtLeast(0)
    @JvmField val lastModified: Long = lastModified.coerceAtLeast(0L)
    @JvmField val fingerprint: String = fingerprint.orEmpty()
    @JvmField val playCount: Int = playCount.coerceAtLeast(0)
    @JvmField val skipCount: Int = skipCount.coerceAtLeast(0)
    @JvmField val dateAdded: Long = dateAdded.coerceAtLeast(0L)
    @JvmField val lastPlayedAt: Long = lastPlayedAt.coerceAtLeast(0L)
    @JvmField val lastCompletedAt: Long = lastCompletedAt.coerceAtLeast(0L)
    @JvmField val normalizedSearchText: String = buildSearchText()

    constructor(uri: String, title: String, artist: String) :
        this(uri, title, artist, "Unknown album", "Unknown genre", 0)

    constructor(uri: String, title: String, artist: String, album: String, genre: String?) :
        this(uri, title, artist, album, genre, 0)

    constructor(
        uri: String,
        title: String,
        artist: String,
        album: String,
        genre: String?,
        durationMs: Int,
    ) : this(
        TrackIdentity.fromLegacyUri(uri),
        uri,
        title,
        artist,
        album,
        genre,
        durationMs,
        -1L,
        0L,
        "",
    )

    constructor(
        trackId: String?,
        uri: String,
        title: String,
        artist: String,
        album: String,
        genre: String?,
        durationMs: Int,
        fileSize: Long,
        lastModified: Long,
        fingerprint: String?,
    ) : this(
        trackId,
        uri,
        title,
        artist,
        album,
        artist,
        genre,
        0,
        0,
        0,
        durationMs,
        fileSize,
        lastModified,
        fingerprint,
        0,
        0,
        System.currentTimeMillis(),
        0L,
        0L,
    )

    fun withLocation(
        newUri: String,
        newSize: Long,
        newLastModified: Long,
        newFingerprint: String?,
    ): Track = Track(
        trackId,
        newUri,
        title,
        artist,
        album,
        genre,
        durationMs,
        newSize,
        newLastModified,
        newFingerprint,
    ).withDetails(this)

    fun withMetadata(
        newTitle: String,
        newArtist: String,
        newAlbum: String,
        newAlbumArtist: String?,
        newGenre: String?,
        newYear: Int,
        newTrackNumber: Int,
        newDiscNumber: Int,
    ): Track = Track(
        trackId,
        uri,
        newTitle,
        newArtist,
        newAlbum,
        newAlbumArtist,
        newGenre,
        newYear,
        newTrackNumber,
        newDiscNumber,
        durationMs,
        fileSize,
        lastModified,
        fingerprint,
        playCount,
        skipCount,
        dateAdded,
        lastPlayedAt,
        lastCompletedAt,
    )

    fun withPlaybackStats(
        newPlayCount: Int,
        newSkipCount: Int,
        newLastPlayedAt: Long,
        newLastCompletedAt: Long,
    ): Track = Track(
        trackId,
        uri,
        title,
        artist,
        album,
        albumArtist,
        genre,
        year,
        trackNumber,
        discNumber,
        durationMs,
        fileSize,
        lastModified,
        fingerprint,
        newPlayCount,
        newSkipCount,
        dateAdded,
        newLastPlayedAt,
        newLastCompletedAt,
    )

    fun asUri(): Uri = Uri.parse(uri)

    private fun withDetails(source: Track): Track = Track(
        trackId,
        uri,
        title,
        artist,
        album,
        source.albumArtist,
        genre,
        source.year,
        source.trackNumber,
        source.discNumber,
        durationMs,
        fileSize,
        lastModified,
        fingerprint,
        source.playCount,
        source.skipCount,
        source.dateAdded,
        source.lastPlayedAt,
        source.lastCompletedAt,
    )

    private fun buildSearchText(): String = buildString {
        append(title).append(' ').append(artist).append(' ').append(album).append(' ').append(genre)
        if (normalizeSearchText(albumArtist) != normalizeSearchText(artist)) {
            append(' ').append(albumArtist)
        }
        if (year > 0) append(' ').append(year)
    }.let(::normalizeSearchText)

    companion object {
        private val whitespace = Regex("\\s+")

        @JvmStatic
        fun normalizeSearchText(value: String?): String = value
            ?.lowercase(Locale.ROOT)
            ?.replace(whitespace, " ")
            ?.trim()
            .orEmpty()
    }
}
